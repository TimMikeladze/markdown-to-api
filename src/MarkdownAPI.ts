import { globbySync } from 'globby';
import matter from 'gray-matter';
import slugify from 'slugify';
import MiniSearch, { Options as MiniSearchOptions } from 'minisearch';
import { remark } from 'remark';
import strip from 'strip-markdown';
import { existsSync, readFileSync, writeFileSync } from 'fs';
import { parse } from 'yaml';
import deepmerge from 'deepmerge';

const toSlug = (string: string): string => slugify(string, { lower: true, strict: true });

export interface FileMetadata {
  description?: string;
  id: string,
  path: string,
  slug?: string;
  tags?: string[];
  title?: string;
}

export interface ParsedFile {
  content: string;
  metadata: FileMetadata;
  strippedContent: string;
}

export interface Options {
  config?: any;
  configPath?: string;
  directory: string;
  indexPath?: string;
  miniSearchOptions?: MiniSearchOptions<any>;
  useIndex?: boolean;
  writeIndex?: boolean;
}

export interface Config {
  fields?: Record<string, {
    required?: boolean;
  } | null>,
  tags?: Record<string, {
    description?: string;
  } | null>
}

const defaultConfig: Config = {
  tags: {},
  fields: {
    title: {
      required: false,
    },
    description: {
      required: false,
    },
    tags: {
      required: false,
    },
  },
};

export class MarkdownAPI {
  private static readonly stripper = remark()
    .use(strip);

  private filePaths!: string[];

  private miniSearch!: MiniSearch;

  private files!: Map<string, ParsedFile>;

  private readonly options: Options;

  private config!: Config;

  constructor(options: Options) {
    this.options = options;
    this.init();
  }

  public getOptions(): Options {
    return this.options;
  }

  public init() {
    this.config = this.loadConfig();
    this.filePaths = this.loadFilePaths();
    this.files = this.loadFiles();
    if (this.options.useIndex && existsSync(this.getIndexPath())) {
      this.miniSearch = this.loadIndexJSON();
    } else {
      this.miniSearch = this.buildIndex();
    }
    if (this.options.writeIndex) {
      this.writeIndexJSON();
    }
  }

  public parseFile(path: string): ParsedFile {
    const output = matter(readFileSync(path, 'utf8'));

    let {
      data: {
        // eslint-disable-next-line prefer-const
        id, title, slug, tags, description,
      },
    }: {
      data: Partial<FileMetadata>;
    } = output;

    if (!title || !title.trim().length) {
      title = path.split('/').pop()?.split('.').shift();
    }

    if (!slug || !slug.trim().length) {
      slug = toSlug(title as string);
    }

    if (!id) {
      const parts = path.split('/');
      id = `${parts.slice(1, parts.length - 1).join('-')}-${slug}`;
    }

    id = toSlug(id);

    if (!description) {
      description = '';
    }

    const defaultFieldKeys = Object.keys(defaultConfig.fields || {});

    const extraFieldKeys = Object.keys(this.config.fields || {})
      .map((x) => (this.config.fields?.[x] && !defaultFieldKeys.includes(x) ? x : null))
      .filter(Boolean) as string[];

    const extraFields = extraFieldKeys.reduce<Record<string, string>>((res, key) => ({
      ...res,
      [key]: (output.data as any)[key],
    }), {});

    const strippedContent = (MarkdownAPI.stripper.processSync(output.content)).toString();

    return {
      metadata: {
        id,
        title,
        slug,
        path,
        tags,
        description,
        ...extraFields,
      },
      content: output.content,
      strippedContent,
    };
  }

  public loadConfig() {
    let config = {};
    if (this.options.configPath) {
      config = parse(readFileSync(this.options.configPath, 'utf8'));
    } else if (this.options.config) {
      config = this.options.config;
    } else if (existsSync(`${this.options.directory}/config.yml`)) {
      config = parse(readFileSync(`${this.options.directory}/config.yml`, 'utf8'));
    } else if (existsSync(`${this.options.directory}/config.yaml`)) {
      config = parse(readFileSync(`${this.options.directory}/config.yaml`, 'utf8'));
    }

    return deepmerge(defaultConfig, config);
  }

  public getConfig(): Config {
    return this.config;
  }

  public getTags(): {
    description?: string;
    name: string;
  }[] {
    return Object.keys(this.config.tags || {}).map((x) => ({
      name: x,
      description: this.config.tags?.[x]?.description || '',
    }));
  }

  public loadFilePaths() {
    const paths = globbySync(this.options.directory, {
      expandDirectories: {
        files: ['*.md'],
      },
    });
    return paths;
  }

  public loadFiles(): Map<string, ParsedFile> {
    const parsedFiles = this.filePaths.map((x) => this.parseFile(x));

    const requiredFields = Object.keys(this.config.fields || {})
      .map((x) => (this.config.fields?.[x]?.required ? x : null))
      .filter(Boolean);

    let errors: string[] = [];

    const files = parsedFiles.reduce((map, file) => {
      errors = [
        ...errors,
        ...requiredFields.reduce<string[]>((res, requiredField) => {
          if (requiredField && !((file.metadata as any)[requiredField])) {
            return [...res, `Missing required field '${requiredField}' in ${file.metadata.path}`];
          }
          return res;
        }, []),
      ];

      map.set(file.metadata.id, file);

      return map;
    }, new Map<string, ParsedFile>());

    if (errors.length) {
      throw new Error(errors.join('\n'));
    }

    return files;
  }

  // eslint-disable-next-line class-methods-use-this
  public getIndexFields() {
    const fields = ['title', 'slug', 'path', 'tags', 'description'];
    return {
      fields: [...fields, 'strippedContent'],
      storeFields: fields,
    };
  }

  public buildIndex(): MiniSearch {
    const miniSearch = new MiniSearch({
      ...this.getIndexFields(),
      ...this.options.miniSearchOptions,
    });

    miniSearch.addAll(Array.from(this.files.values()).map((x) => ({ ...x.metadata, ...x })));

    return miniSearch;
  }

  public getFile(id: string): ParsedFile | undefined {
    return this.files.get(id);
  }

  public getFiles(): Map<string, ParsedFile> {
    return this.files;
  }

  public getIndex() {
    return this.miniSearch;
  }

  public indexToJSON(): any {
    return JSON.stringify(this.miniSearch.toJSON());
  }

  public writeIndexJSON(): void {
    writeFileSync(this.getIndexPath(), this.indexToJSON(), 'utf8');
  }

  public getIndexPath(): string {
    return this.options.indexPath || `${this.options.directory}/index.json`;
  }

  public loadIndexJSON() {
    if (this.getIndexPath()) {
      return MiniSearch.loadJSON(readFileSync(this.getIndexPath(), 'utf8'), {
        ...this.getIndexFields(),
        ...this.options.miniSearchOptions,
      });
    }
    throw new Error('No index path provided');
  }

  public getFilePaths(): string[] {
    return this.filePaths;
  }
}
