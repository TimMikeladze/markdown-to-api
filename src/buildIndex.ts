import { globby } from 'globby';
import { readFile } from 'fs/promises';
import matter from 'gray-matter';
import slugify from 'slugify';
import MiniSearch from 'minisearch';
import { remark } from 'remark';
import strip from 'strip-markdown';

export const loadFilePaths = async (path: string) => {
  const paths = await globby(path, {
    expandDirectories: {
      files: ['*.md'],
    },
  });
  return paths;
};

export interface FileData {
  id: string,
  path: string,
  slug?: string;
  tags?: string[];
  title?: string;
}

export interface ParsedFile {
  content: string;
  data: FileData;
  rawContent: string;
}

const stripper = remark()
  .use(strip);

export const parseFile = async (path: string): Promise<ParsedFile> => {
  const output = matter(await readFile(path, 'utf8'));

  let {
    data: {
      // eslint-disable-next-line prefer-const
      id, title, slug, tags,
    },
  }: {
    data: Partial<FileData>;
  } = output;

  if (!title || !title.trim().length) {
    title = path.split('/').pop()?.split('.').shift();
  }

  if (!slug || !slug.trim().length) {
    slug = slugify(title as string, {
      lower: true,
      strict: true,
    });
  }

  if (!id) {
    const parts = path.split('/');
    id = `${parts.slice(0, parts.length - 1).join('/')}/${slug}`;
  }

  const rawContent = (await stripper.process(output.content)).toString();

  return {
    data: {
      id,
      title,
      slug,
      path,
      tags,
    },
    content: output.content,
    rawContent,
  };
};

export const parseFiles = (paths: string[]): Promise<ParsedFile[]> => {
  const files = Promise.all(paths.map(parseFile));
  return files;
};

export const buildIndex = async (files: ParsedFile[]) => {
  const fields = ['title', 'slug', 'path', 'tags'];

  const miniSearch = new MiniSearch({
    fields: [...fields, 'content'],
    storeFields: fields,
  });

  await miniSearch.addAllAsync(files.map((x) => ({ ...x.data, content: x.content })));

  return miniSearch;
};
