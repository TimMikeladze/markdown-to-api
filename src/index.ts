import { globby } from 'globby';
import { readFile } from 'fs/promises';
import matter from 'gray-matter';
import slugify from 'slugify';
import MiniSearch from 'minisearch';
import { remark } from 'remark';
import strip from 'strip-markdown';
import { existsSync, readFileSync } from 'fs';
import { parse } from 'yaml';
import deepmerge from 'deepmerge';

export interface FileData {
  description?: string;
  id: string,
  path: string,
  slug?: string;
  tags?: string[];
  title?: string;
}

export interface ParsedFile {
  content: string;
  data: FileData;
  strippedContent: string;
}

export interface Options {
  config?: any;
  configPath?: string;
  directory: string;
}

export interface Config {
  fields?: Record<string, {
    required?: boolean;
  } | null>,
  tags?: Record<string, {
    description?: string;
  } | null>
}

const defaultConfig: Config = {
  tags: {},
  fields: {
    title: {
      required: false,
    },
    description: {
      required: false,
    },
    tags: {
      required: false,
    },
  },
};

export class MarkdownToGraphQL {
  private static readonly stripper = remark()
    .use(strip);

  private filePaths!: string[];

  private miniSearch!: MiniSearch;

  private files!: Map<string, ParsedFile>;

  private readonly options: Options;

  private readonly config: Config;

  constructor(options: Options) {
    this.options = options;
    this.config = this.loadConfig();
  }

  public async init() {
    this.filePaths = await this.loadFilePaths();
    this.files = await this.loadFiles();
    this.miniSearch = await this.buildIndex();
  }

  public loadConfig() {
    let config = {};
    if (this.options.configPath) {
      config = parse(readFileSync(this.options.configPath, 'utf8'));
    } else if (this.options.config) {
      config = this.options.config;
    } else if (existsSync(`${this.options.directory}/config.yml`)) {
      config = parse(readFileSync(`${this.options.directory}/config.yml`, 'utf8'));
    } else if (existsSync(`${this.options.directory}/config.yaml`)) {
      config = parse(readFileSync(`${this.options.directory}/config.yaml`, 'utf8'));
    }

    return deepmerge(defaultConfig, config);
  }

  public getConfig(): Config {
    return this.config;
  }

  public static async parseFile(path: string): Promise<ParsedFile> {
    const output = matter(await readFile(path, 'utf8'));

    let {
      data: {
        // eslint-disable-next-line prefer-const
        id, title, slug, tags, description,
      },
    }: {
      data: Partial<FileData>;
    } = output;

    if (!title || !title.trim().length) {
      title = path.split('/').pop()?.split('.').shift();
    }

    if (!slug || !slug.trim().length) {
      slug = slugify(title as string, {
        lower: true,
        strict: true,
      });
    }

    if (!id) {
      const parts = path.split('/');
      id = `${parts.slice(0, parts.length - 1).join('/')}/${slug}`;
    }

    const strippedContent = (await MarkdownToGraphQL.stripper.process(output.content)).toString();

    return {
      data: {
        id,
        title,
        slug,
        path,
        tags,
        description,
      },
      content: output.content,
      strippedContent,
    };
  }

  public async loadFilePaths() {
    const paths = await globby(this.options.directory, {
      expandDirectories: {
        files: ['*.md'],
      },
    });
    return paths;
  }

  public async loadFiles(): Promise<Map<string, ParsedFile>> {
    const parsedFiles = await Promise.all(this.filePaths.map(MarkdownToGraphQL.parseFile));

    const requiredFields = Object.keys(this.config.fields || {}).map((x) => (this.config.fields?.[x]?.required ? x : null)).filter(Boolean);

    let errors: string[] = [];

    const files = parsedFiles.reduce((map, file) => {
      errors = [
        ...errors,
        ...requiredFields.reduce<string[]>((res, requiredField) => {
          if (requiredField && !((file.data as any)[requiredField])) {
            return [...res, `Missing required field '${requiredField}' in ${file.data.path}`];
          }
          return res;
        }, []),
      ];

      map.set(file.data.id, file);

      return map;
    }, new Map<string, ParsedFile>());

    if (errors.length) {
      throw new Error(errors.join('\n'));
    }

    return files;
  }

  public async buildIndex(): Promise<MiniSearch> {
    const fields = ['title', 'slug', 'path', 'tags', 'description'];

    const miniSearch = new MiniSearch({
      fields: [...fields, 'strippedContent'],
      storeFields: fields,
    });

    await miniSearch.addAllAsync(Array.from(this.files.values()).map((x) => ({ ...x.data, content: x.content })));

    return miniSearch;
  }

  public getFile(id: string): ParsedFile | undefined {
    return this.files.get(id);
  }

  public getFiles(): Map<string, ParsedFile> {
    return this.files;
  }

  public getIndex() {
    return this.miniSearch;
  }

  public getFilePaths(): string[] {
    return this.filePaths;
  }
}
