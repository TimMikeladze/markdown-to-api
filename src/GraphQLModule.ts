import { createModule, gql } from 'graphql-modules';
import { SearchOptions } from 'minisearch';
import { MarkdownAPI, Options } from './MarkdownAPI';

export const typeDefs = gql`
  type Query {
    markdownFile(id: ID!): MarkdownFile!
    searchMarkdownFiles(text: String! options: MarkdownFileSearchOptions): MarkdownFileSearchResults!
  }

  type MarkdownFile {
    id: ID!
    path: String!
    slug: String!
    tags: [String!]
    title: String!
    description: String
    content: String!
    strippedContent: String!
  }

  type MarkdownFileSearchResults {
    count: Int!
    results: [MarkdownFileSearchResult!]!
  }

  type MarkdownFileSearchResult {
    id: ID!
    description: String
    score: Float!
    terms: [String!]
    title: String!
    slug: String!
    path: String!
    tags: [String!]
    markdownFile: MarkdownFile!
  }

  input MarkdownFileSearchOptions {
    fields: [String!]
    weights: MarkdownFileSearchOptionsWeights
    prefix: Boolean
    fuzzy: Boolean
    maxFuzzy: Int
    combineWith: MarkdownFileSearchOptionsCombineWith
  }

  enum MarkdownFileSearchOptionsCombineWith {
    AND
    OR
  }

  input MarkdownFileSearchOptionsWeights {
    fuzzy: Float!
    exact: Float!
  }
`;

export const getResolvers = (mdapi: MarkdownAPI) => ({
  Query: {
    markdownFile(_: any, { id }: { id: string }) {
      const file = mdapi.getFile(id);
      if (!file) {
        throw new Error(`File not found: ${id}`);
      }
      return {
        ...file,
        ...file.metadata,
      };
    },
    searchMarkdownFiles(_: any, { text, options }: { options: SearchOptions, text: string }) {
      const results = mdapi.getIndex().search(text, {
        ...mdapi.getOptions(),
        ...options,
      });
      return {
        count: results.length,
        results,
      };
    },
  },
  MarkdownFileSearchResult: {
    markdownFile({ id }: { id: string }) {
      if (!id) {
        throw new Error('selection set is missing id');
      }
      const file = mdapi.getFile(id);
      return {
        ...file,
        ...file!.metadata,
      };
    },
  },
});

export const createMarkdownAPIModule = (options: Options) => {
  const mdapi = new MarkdownAPI(options);

  return createModule({
    id: 'MarkdownAPIModule',
    typeDefs,
    resolvers: getResolvers(mdapi),
  });
};
