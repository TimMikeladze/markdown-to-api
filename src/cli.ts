import { Command } from 'commander';
import { MarkdownAPI } from './MarkdownAPI';

const program = new Command();

program
  .name('markdown-to-api')
  .description('CLI to interact with markdown-to-api');

program.requiredOption(
  '-d --directory <folder>',
  'Path to the folder containing the markdown files',
);

program.option(
  '-c --configPath <configPath>',
  'Path to the config file',
);

program.option(
  '-i --indexPath <indexPath>',
  'Path to the index file',
);

try {
  const res = program.parse(process.argv);

  const mdapi = new MarkdownAPI(res.opts());

  mdapi.writeIndexJSON();

  // eslint-disable-next-line no-console
  console.log(`Search index written to: ${mdapi.getIndexPath()}`);
} catch (err: any) {
  // eslint-disable-next-line no-console
  console.error(err.message);
}
