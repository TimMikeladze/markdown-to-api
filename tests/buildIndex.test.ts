import { MarkdownToGraphQL } from '../src';

describe('buildIndex', () => {
  it('builds an index', async () => {
    const mg = new MarkdownToGraphQL({
      directory: 'tests/basic',
    });
    await mg.init();
    expect(mg.getIndex()).toMatchSnapshot();
  });
});
