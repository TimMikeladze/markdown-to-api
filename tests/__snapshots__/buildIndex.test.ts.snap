// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`buildIndex builds an index 1`] = `
Object {
  "averageFieldLength": Array [
    2.142857142857143,
    2.2857142857142856,
    5.142857142857143,
    0.3333333333333333,
    1,
  ],
  "documentCount": 7,
  "documentIds": Object {
    "0": "tests/basic/hello-world",
    "1": "tests/basic/no-slug-or-title",
    "2": "tests/basic/no-slug",
    "3": "tests/basic/no-title",
    "4": "tests/basic/with-description",
    "5": "tests/basic/with-tags",
    "6": "tests/basic/nested/nested-filed",
  },
  "fieldIds": Object {
    "description": 4,
    "path": 2,
    "slug": 1,
    "strippedContent": 5,
    "tags": 3,
    "title": 0,
  },
  "fieldLength": Object {
    "0": Array [
      1,
      2,
      4,
    ],
    "1": Array [
      4,
      4,
      7,
    ],
    "2": Array [
      2,
      2,
      5,
    ],
    "3": Array [
      2,
      2,
      5,
    ],
    "4": Array [
      2,
      2,
      5,
      ,
      5,
    ],
    "5": Array [
      2,
      2,
      5,
      2,
    ],
    "6": Array [
      2,
      2,
      5,
    ],
  },
  "index": Array [
    Array [
      "filed",
      Object {
        "1": Object {
          "6": 1,
        },
      },
    ],
    Array [
      "file",
      Object {
        "0": Object {
          "6": 1,
        },
        "2": Object {
          "6": 1,
        },
      },
    ],
    Array [
      "nested",
      Object {
        "0": Object {
          "6": 1,
        },
        "1": Object {
          "6": 1,
        },
        "2": Object {
          "6": 2,
        },
      },
    ],
    Array [
      "no",
      Object {
        "0": Object {
          "1": 1,
          "2": 1,
          "3": 1,
        },
        "1": Object {
          "1": 1,
          "2": 1,
          "3": 1,
        },
        "2": Object {
          "1": 1,
          "2": 1,
          "3": 1,
        },
      },
    ],
    Array [
      "a",
      Object {
        "4": Object {
          "4": 1,
        },
      },
    ],
    Array [
      "is",
      Object {
        "4": Object {
          "4": 1,
        },
      },
    ],
    Array [
      "description",
      Object {
        "0": Object {
          "4": 1,
        },
        "1": Object {
          "4": 1,
        },
        "2": Object {
          "4": 1,
        },
        "4": Object {
          "4": 1,
        },
      },
    ],
    Array [
      "with",
      Object {
        "0": Object {
          "4": 1,
          "5": 1,
        },
        "1": Object {
          "4": 1,
          "5": 1,
        },
        "2": Object {
          "4": 1,
          "5": 1,
        },
      },
    ],
    Array [
      "world",
      Object {
        "1": Object {
          "0": 1,
        },
      },
    ],
    Array [
      "tag2",
      Object {
        "3": Object {
          "5": 1,
        },
      },
    ],
    Array [
      "tag1",
      Object {
        "3": Object {
          "5": 1,
        },
      },
    ],
    Array [
      "tags",
      Object {
        "0": Object {
          "5": 1,
        },
        "1": Object {
          "5": 1,
        },
        "2": Object {
          "5": 1,
        },
      },
    ],
    Array [
      "this",
      Object {
        "4": Object {
          "4": 1,
        },
      },
    ],
    Array [
      "title",
      Object {
        "0": Object {
          "1": 1,
          "3": 1,
        },
        "1": Object {
          "1": 1,
          "3": 1,
        },
        "2": Object {
          "1": 1,
          "3": 1,
        },
      },
    ],
    Array [
      "tests",
      Object {
        "2": Object {
          "0": 1,
          "1": 1,
          "2": 1,
          "3": 1,
          "4": 1,
          "5": 1,
          "6": 1,
        },
      },
    ],
    Array [
      "or",
      Object {
        "0": Object {
          "1": 1,
        },
        "1": Object {
          "1": 1,
        },
        "2": Object {
          "1": 1,
        },
      },
    ],
    Array [
      "slug",
      Object {
        "0": Object {
          "1": 1,
          "2": 1,
        },
        "1": Object {
          "1": 1,
          "2": 1,
        },
        "2": Object {
          "1": 1,
          "2": 1,
        },
      },
    ],
    Array [
      "md",
      Object {
        "2": Object {
          "0": 1,
          "1": 1,
          "2": 1,
          "3": 1,
          "4": 1,
          "5": 1,
          "6": 1,
        },
      },
    ],
    Array [
      "basic",
      Object {
        "2": Object {
          "0": 1,
          "1": 1,
          "2": 1,
          "3": 1,
          "4": 1,
          "5": 1,
          "6": 1,
        },
      },
    ],
    Array [
      "hello",
      Object {
        "0": Object {
          "0": 1,
        },
        "1": Object {
          "0": 1,
        },
        "2": Object {
          "0": 1,
        },
      },
    ],
  ],
  "nextId": 7,
  "serializationVersion": 2,
  "storedFields": Object {
    "0": Object {
      "path": "tests/basic/hello.md",
      "slug": "hello-world",
      "title": "Hello",
    },
    "1": Object {
      "path": "tests/basic/no-slug-or-title.md",
      "slug": "no-slug-or-title",
      "title": "no-slug-or-title",
    },
    "2": Object {
      "path": "tests/basic/no-slug.md",
      "slug": "no-slug",
      "title": "No Slug",
    },
    "3": Object {
      "path": "tests/basic/no-title.md",
      "slug": "no-title",
      "title": "no-title",
    },
    "4": Object {
      "description": "This is a description.",
      "path": "tests/basic/with-description.md",
      "slug": "with-description",
      "title": "With description",
    },
    "5": Object {
      "path": "tests/basic/with-tags.md",
      "slug": "with-tags",
      "tags": Array [
        "tag1",
        "tag2",
      ],
      "title": "With tags",
    },
    "6": Object {
      "path": "tests/basic/nested/nested-file.md",
      "slug": "nested-filed",
      "title": "Nested File",
    },
  },
}
`;
